core/src/main/scala/quasar/destination/gbq/GBQDestination.scala

```diff
@@ -0,0 +1,244 @@
+        _ <- eitherloc match {
+            case Right(locationUri) => {
+              upload(client, bytes, locationUri)
+            }
+            case Left(e) =>
+              Stream.eval(Sync[F].delay(ApplicativeError[F, Throwable].raiseError(
+                new RuntimeException(s"No Location URL from returned from job: $e"))))
+          }
+      } yield ()
+  }
+
+  // NB: config overwrites when re-pushing tables
+  // see schemaUpdateOptions section at
+  // https://cloud.google.com/bigquery/docs/reference/rest/v2/Job
+  private def formGBQJobConfig(
+      schema: NonEmptyList[GBQSchema],
+      config: GBQConfig,
+      tableId: String)
+      : GBQJobConfig =
+    GBQJobConfig(
+      "CSV",
+      "1",
+      "true",
+      schema.toList,
+      Some("DAY"), //TODO: make this configurable
+      WriteDisposition("WRITE_TRUNCATE"), //TODO: make this configurable
+      GBQDestinationTable(config.authCfg.projectId, config.datasetId, tableId), 
+      "21600000", //6hrs load job limit
+      "LOAD")
+
+  def mkErrorString(errs: NonEmptyList[ColumnType.Scalar]): String =
+    errs
+      .map(err => s"Column of type ${err.show} is not supported by Avalanche")
+      .intercalate(", ")
+
+  private def ensureValidColumns(columns: NonEmptyList[Column[ColumnType.Scalar]]): Either[String, NonEmptyList[GBQSchema]] =
+    columns.traverse(mkColumn(_)).toEither leftMap { errs =>
+      s"Some column types are not supported: ${mkErrorString(errs)}"
+    }
+
+  private def mkColumn(c: Column[ColumnType.Scalar]): ValidatedNel[ColumnType.Scalar, GBQSchema] =
+    tblColumnToGbq(c.tpe).map(s => GBQSchema(s, c.name))
+
+  private def tblColumnToGbq(ct: ColumnType.Scalar): ValidatedNel[ColumnType.Scalar, String] =
+    ct match {
+        case ColumnType.String => "STRING".validNel
+        case ColumnType.Boolean => "BOOL".validNel
+        case ColumnType.Number => "NUMERIC".validNel
+        case ColumnType.LocalDate => "DATE".validNel
+        case ColumnType.LocalDateTime => "DATETIME".validNel
+        case ColumnType.LocalTime => "TIME".validNel
+        case ColumnType.OffsetTime => "TIMESTAMP".validNel
+        case ColumnType.OffsetDateTime => "TIMESTAMP".validNel
+        case ColumnType.OffsetDate => "TIMESTAMP".validNel
+        case i @ ColumnType.Interval => i.invalidNel
+        case ColumnType.Null => "INTEGER1".validNel
+      }
+
+  private def mkDataset(client: Client[F], accessToken: String, config: GBQConfig): F[Either[InitializationError[Json], Unit]] = {
+    implicit def jobConfigEntityEncoder: EntityEncoder[F, GBQDatasetConfig] = 
+      jsonEncoderOf[F, GBQDatasetConfig]
+
+    val dCfg = GBQDatasetConfig(config.authCfg.projectId, config.datasetId)
+    val bearerToken = Authorization(Credentials.Token(AuthScheme.Bearer, accessToken))
+    val datasetReq = Request[F](
+      method = Method.POST,
+      uri = Uri.fromString(s"https://bigquery.googleapis.com/bigquery/v2/projects/${config.authCfg.projectId}/datasets")
```

Comment:
@cryogenian is your concern that, if `config.authCfg.projectId` is not a valid uri part, we end up in the `Uri()` case below. In turn, this means that we're handling this error indirectly, i.e. the error will surface as a "whatever happens when the `Uri` is empty" error, instead of a "this failed because your project id is invalid" configuration error?

This explanation makes sense to me, at least :)

Our goal error type should be `DestinationError.InvalidConfiguration`. We don't have to actually run the request to generate this error. One way to do this:

```scala
 private def mkDataset(client: Client[F], accessToken: String, config: GBQConfig): F[Either[InitializationError[Json], Unit]] = {
   
   val uri = Uri.fromString(s"https://bigquery.googleapis.com/bigquery/v2/projects/${config.authCfg.projectId}/datasets")

    uri.toOption match {
      case None => DestinationError.invalidConfiguration(...).asLeft.pure[F] // give the reason as invalid uri with the string
      case Some(validUri) => ... // everything that's in the function now
    }
```
Or instead of `toOption`, we could keep as `Either[ParseFailure, A]` and use the `ParseFailure` message - I haven't checked if that has a useful message but it probably does.

There's also the open question if we should move all this logic into the config decoder. Here, we have it as a runtime configuration error. In the decoder, we'd have it as a static configuration error. It feels wrong putting it in the decoder itself, because it's an implementation detail that we use the `projectId` in a url; we can imagine another GBQ destination implementation that allows `projectId`s to be not valid when used in a url. Though, maybe this is an actual GBQ constraint, and not just a consequent of our implementation?

Reply:
RIght, I meant that we should check this `projectId` during decoding the config :+1: for moving generation errors there

Reply:
👍 to checking this when we decode the config.

Reply:
I'll be using the pattern mentioned above along with passing `config.authCfg.projectId` through `URLEncoder`.